nucleus/payara-modules/requesttracing-core/src/main/java/fish/payara/nucleus/requesttracing/store/strategy/ReservoirTraceStorageStrategy.java

```diff
@@ -51,66 +52,42 @@
 public class ReservoirTraceStorageStrategy implements TraceStorageStrategy {
 
     private final Random random;
-    private long counter;
 
     public ReservoirTraceStorageStrategy() {
         this.random = new Random();
-        this.counter = 0;
     }
 
     /**
-     * Gets the trace that needs removing. Uses a reservoir sampling algorithm
-     * to determine this.
+     * Gets the trace that needs removing. Removes the provided trace if present, or a random trace each trace having
+     * the same probability of being kept or removed.
      *
-     * @param traces the list of traces to test.
-     * @param maxSize the maximum size of the list.
-     * @return the trace that needs removing, or null if no traces need
-     * removing.
-     */
-    @Override
-    public RequestTrace getTraceForRemoval(Collection<RequestTrace> traces, int maxSize) {
-        if (counter < Long.MAX_VALUE) {
-            counter++;
-        }
-
-        // Probability of keeping the new item
-        double probability = (double) maxSize / counter;
-        boolean keepItem = random.nextDouble() < probability;
-
-        // If the probability fails or the list isn't full, nothing should be removed
-        if (!keepItem || traces.size() <= maxSize) {
```

Comment:
Here it should have been `keepItem` instead of `!keepItem` except that other errors in the algorithm partly cancelled this out or shadowed it. Now it does not play a role as we see below.